flfm/app/main.py

```diff
@@ -0,0 +1,470 @@
+                                children=html.Div(id=dict(type="display", index="reconstruction")),
+                            ),
+                        ],
+                    ),
+                ]
+            ),
+        ]
+    )
+    return layout
+
+
+dash_app.layout = dash_layout
+
+
+@callback(
+    Output(dict(type="display", index=MATCH), "children", allow_duplicate=True),
+    Output(dict(type="n-frames", index=MATCH), "data"),
+    Input(dict(type="upload", index=MATCH), "contents"),
+    State(dict(type="upload", index=MATCH), "filename"),
+    State(dict(type="upload", index=MATCH), "id"),
+    prevent_initial_call=True,
+)
+def upload_data(contents, filename, id):
+    """Upload data/image files."""
+    global image_data
+    id = id["index"]
+    image_data[id] = upload_file(contents, filename)
+    children, n_frames_from_image = _display_image_with_slider(image_data[id], id=id)
+    return children, n_frames_from_image
+
+
+@callback(
+    Output("normalize-psf", "disabled", allow_duplicate=True),
+    Input(dict(type="upload", index="psf"), "last_modified"),
+    prevent_initial_call=True,
+)
+def enable_normalization_button(_):
+    """Re-enable normalization button upon new upload."""
+    return False
+
+
+@callback(
+    Output(dict(type="image-graph", index=MATCH), "figure"),
+    Output(dict(type="image-slider", index=MATCH), "tooltip"),
+    Input(dict(type="image-slider", index=MATCH), "value"),
+    State(dict(type="image-slider", index=MATCH), "tooltip"),
+    State(dict(type="n-frames", index="psf"), "data"),  # NOTE: This intentionally uses index="psf".
+    State("depth-step-size", "value"),
+    State(dict(type="color-scale", index=MATCH), "value"),
+    State(dict(type="image-slider", index=MATCH), "id"),
+    prevent_initial_call=True,
+)
+def update_image_from_slider(value, tooltip, n_frames, step_size, color_scale, id):
+    """Re-plot image based on frame/depth selected from slider."""
+    global image_data
+    # tooltip.transform (clientside js )would render faster than updating the template in a server side callback,
+    # however, it only takes a single input ``value`` and there doesn't seem a trivial way to also pass ``step_size``
+    # and ``n_frames`` to the js transform function (and have it dynamically served/reloaded).
+    tooltip["template"] = slider_template_str(value, _center_frame(n_frames), step_size)
+    fig = plot_image(image_data[id["index"]][value, :, :], color_scale=color_scale)
+    return fig, tooltip
+
+
+@callback(
+    Output({"type": "image-slider", "index": ALL}, "tooltip", allow_duplicate=True),
+    Input("depth-step-size", "value"),
+    State({"type": "image-slider", "index": ALL}, "value"),
+    State({"type": "image-slider", "index": ALL}, "tooltip"),
+    State({"type": "n-frames", "index": "psf"}, "data"),
+    prevent_initial_call=True,
+)
+def update_slider_tooltip(depth_step_size, values, tooltips, n_frames):
```

Comment:
Can you add some type hints here if the types are consistent and reasonable to reference?